--- setup.py.old	2009-07-13 11:36:43.375000000 -0500
+++ setup.py	2009-07-13 11:53:28.000000000 -0500
@@ -64,7 +64,10 @@
     # their cURL source installation.  The path set here is just an example
     # and thus unlikely to match your installation.
     CURL_DIR = r"c:\src\build\pycurl\curl-7.16.2.1"
+    if os.environ.has_key("CURL_DIR"):
+        CURL_DIR = os.environ["CURL_DIR"]
     CURL_DIR = scan_argv("--curl-dir=", CURL_DIR)
+    
     print "Using curl directory:", CURL_DIR
     assert os.path.isdir(CURL_DIR), "please check CURL_DIR in setup.py"
     include_dirs.append(os.path.join(CURL_DIR, "include"))
