Index: cherrypy/wsgiserver/__init__.py
===================================================================
--- cherrypy/wsgiserver/__init__.py	(revision 2482)
+++ cherrypy/wsgiserver/__init__.py	(working copy)
@@ -1570,6 +1570,9 @@
         if isinstance(self.bind_addr, basestring):
             # AF_UNIX socket
             
+            # AF_UNIX socket doesn't have TCP options
+            self.nodelay = False
+            
             # So we can reuse the socket...
             try: os.unlink(self.bind_addr)
             except: pass
