Index: cherrypy/_cpwsgi.py
===================================================================
--- cherrypy/_cpwsgi.py	2009-04-12 22:38:03.000000000 -0700
+++ cherrypy/_cpwsgi.py	2010-05-18 12:59:35.402733878 -0700
@@ -220,15 +220,22 @@
         return response.status, response.header_list, response.body
     
     def get_request(self):
         """Create a Request object using environ."""
         env = self.environ.get
+
+        try:
+                local_port = int(env('SERVER_PORT', 80))
+        except ValueError:
+                local_port = -1
+        try:
+                remote_port = int(env('REMOTE_PORT', -1))
+        except ValueError:
+                remote_port = -1
         
-        local = _http.Host('', int(env('SERVER_PORT', 80)),
-                           env('SERVER_NAME', ''))
-        remote = _http.Host(env('REMOTE_ADDR', ''),
-                            int(env('REMOTE_PORT', -1)),
+        local = _http.Host('', local_port, env('SERVER_NAME', ''))
+        remote = _http.Host(env('REMOTE_ADDR', ''), remote_port,
                             env('REMOTE_HOST', ''))
         scheme = env('wsgi.url_scheme')
         sproto = env('ACTUAL_SERVER_PROTOCOL', "HTTP/1.1")
         request, resp = self.cpapp.get_serving(local, remote, scheme, sproto)
         
