--- setup.py	2008-04-11 10:53:24.000000000 -0500
+++ setup.py.new	2009-08-06 17:16:03.093750000 -0500
@@ -55,8 +55,8 @@
 
 # Add more platforms here when needed
 if os.name == 'nt' or sys.platform == 'win32':
-    Libraries = ['eay32', 'Ws2_32']
-    ExtraObjects = [r"c:\Python25\libs\ssleay32.a"]
+    Libraries = ['libeay32', 'ssleay32', 'Ws2_32']
+    ExtraObjects = []
 else:
     Libraries = ['ssl', 'crypto']
     ExtraObjects = []
